firmware/src/networking/wifi/esp.rs

```diff
@@ -40,56 +61,85 @@ pub async fn network_task(signals: &Signals) {
 		&mut tx_buffer,
 	);
 
+	// Server will send broadcasts to this port
 	socket.bind(PORT).unwrap();
 
-	let mut buffer = [0u8; 256];
+	// Sequence numbers are monotonically increasing. This is done to reject out-of-order packets
+	// This along with serialization should maybe be done in Packets
+	let mut tx_seq = 0;
+	let mut rx_seq = 0;
+
+	// TODO: Implement with proper async select. So far there is no async counterpart of recv
 	loop {
-		main_loop(&mut socket, signals, &mut buffer).await;
-	}
-}
+		// Either start sending or receive, if either is available
+		let net = select(
+			packets.serverbound.recv(),
+			recv_bytes(&mut socket, &mut buffer),
```

Comment:
whichever future is slower will dropped, often referred to as async cancellation. Any internal buffers in it will be dropped too, causing unobserved data loss - a cause of really gnarly and subtle bugs.

With how simple the code is right now, this is probably a non-issue. But it could be an issue as the code evolves, and I would rather just avoid it entirely. lets avoid any possible async cancellation issues and just structure this as two futures - each future does a `loop` and for every item it encounters, it will send it (or vice versa).

These two futures can be combined into one with `embassy_futures::join()`. We then simply `.await` on this joined future. Because the two inner futures both have `loop`s, this outer future will never return.

This also implies that the outer `loop` on line 73 would disappear, since awaiting the joined future will never return.

All of the above is a nontrivial change, so I think I'll accept this as-is right now, and fix that shortly, to keep the review burden of PRs low.

Reply:
Future cancellations are tricky, I agree. However, here I used select because I wanted to share `buffer` for both receiving and sending, reducing the memory needed. Sharing a `&mut [u8]` between two futures is even more non-trivial :)

Ideally, esp-wifi would expose the internal `tx_buffer` and `rx_buffer`s, so we could just directly operate on-place with them and not need `buffer`. smoltcp already does this, just limitation in esp-wifi API surface

Reply:
Good point about sharing the buffer, I didn't think of that. Still not sure if its worth it. Like I mentioned, I'll accept it as is for now.

`esp-wifi` would be likely to accept a contribution that exposes their buffers, assuming that that can be done in rust and wouldn't require modification to their binary c libs.